### PR TITLE
refactor!: rework bulk handler validation 

### PR DESCRIPTION
**Pull request by kleyow**
Rework bulk handler validator `reasons` string array to a `FSPIOPError` array to better specify error codes and descriptions.
Follow up PR for mojaloop/project#2800 mojaloop/project#2798

Test cases updated here https://github.com/mojaloop/testing-toolkit-test-cases/pull/79

Breaking changes:
The FSPIOP api errors for invalid/inactive/non-existant FSP's for **Bulk Transfers** are now updated to be more descriptive.
They have been updated to 
- ErrorHandler.Enums.FSPIOPErrorCodes.PAYER_FSP_ID_NOT_FOUND - 3202
- ErrorHandler.Enums.FSPIOPErrorCodes.PAYEE_FSP_ID_NOT_FOUND - 3203

If you have conditional logic based on error callbacks of POST /bulkTransfers you will need to update the error codes accordingly.

### CODE REVIEW

File: src/handlers/bulk/shared/validator.js [hunk 1]
```diff
@@ -37,28 +37,44 @@
...
 const reasons = []
 
 const validateDifferentFsp = (payload) => {
   const isPayerAndPayeeDifferent = (payload.payerFsp.toLowerCase() !== payload.payeeFsp.toLowerCase())
   if (!isPayerAndPayeeDifferent) {
-    reasons.push('Payer and Payee should differ')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'Payer and Payee should differ'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:50]:
```suggestion
        'Payer and Payee FSPs should be different'
```

File: src/handlers/bulk/shared/validator.js [hunk 2]
```diff
@@ -37,28 +37,44 @@
...
   return true
 }
 const validateFspiopSourceMatchesPayer = (payload, headers) => {
   const matched = (headers && headers[Enum.Http.Headers.FSPIOP.SOURCE] === payload.payerFsp)
   if (!matched) {
-    reasons.push('FSPIOP-Source header should match Payer')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'FSPIOP-Source header should match Payer'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:75]:
```suggestion
        'FSPIOP-Source header should match Payer FSP'
```

File: src/handlers/bulk/shared/validator.js [hunk 3]
```diff
@@ -68,24 +84,53 @@ const validateFspiopSourceAndDestination = async (payload, headers) => {
   const matchedPayee = (headers && headers[Enum.Http.Headers.FSPIOP.SOURCE] === participant.payeeFsp)
   const matchedPayer = (headers && headers[Enum.Http.Headers.FSPIOP.DESTINATION] === participant.payerFsp)
   if (!matchedPayee) {
-    reasons.push('FSPIOP-Source header should match Payee')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'FSPIOP-Source header should match Payee'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:90]:
```suggestion
        'FSPIOP-Source header should match Payee FSP'
```

File: src/handlers/bulk/shared/validator.js [hunk 4]
```diff
@@ -68,24 +84,53 @@ const validateFspiopSourceAndDestination = async (payload, headers) => {
...
+      )
+    )
     return false
   }
   if (!matchedPayer) {
-    reasons.push('FSPIOP-Destination header should match Payer')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'FSPIOP-Destination header should match Payer'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:99]:
```suggestion
        'FSPIOP-Destination header should match Payer FSP'
```

File: src/handlers/bulk/shared/validator.js [hunk 5]
```diff
@@ -99,15 +144,20 @@ const validateBulkTransfer = async (payload, headers) => {
   let payerParticipantId = null
   let payeeParticipantId = null
   if (!payload) {
-    reasons.push('Bulk transfer must be provided')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'Bulk transfer must be provided'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:150]:
```suggestion
        'A valid Bulk transfer message must be provided'
```

File: src/handlers/bulk/shared/validator.js [hunk 6]
```diff
@@ -132,7 +189,12 @@ const validateBulkTransferFulfilment = async (payload, headers) => {
   reasons.length = 0
   let isValid = true
   if (!payload) {
-    reasons.push('Bulk transfer fulfilment payload must be provided')
+    reasons.push(
+      ErrorHandler.Factory.createFSPIOPError(
+        ErrorHandler.Enums.FSPIOPErrorCodes.VALIDATION_ERROR,
+        'Bulk transfer fulfilment payload must be provided'
```
Comment (elnyry-sam-k) [inline src/handlers/bulk/shared/validator.js:195]:
```suggestion
        'Valid Bulk transfer fulfilment message must be provided'
```

**Comment by elnyry-sam-k** [comment]
Thanks Kevin, looks good..

just some suggestions on changes to error descriptions / messages..

Nice to capture additional errors as information in extension lists and adding new (specific) validation errors..

**Comment by kleyow** [comment]
@elnyry-sam-k done

**Review by elnyry-sam-k — APPROVED** [review]

**Pull request merged**